vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+}
+
+RRGraph::edge_range RRGraph::node_in_edges(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    return vtr::make_range(node_in_edges_[node].begin(), node_in_edges_[node].end());
+}
+
+//Edge attributes
+size_t RRGraph::edge_index(RREdgeId edge) const {
+    VTR_ASSERT_SAFE(valid_edge_id(edge));
+    return size_t(edge);
+}
+
+RRNodeId RRGraph::edge_src_node(RREdgeId edge) const {
+    VTR_ASSERT_SAFE(valid_edge_id(edge));
+    return edge_src_nodes_[edge];
+}
+RRNodeId RRGraph::edge_sink_node(RREdgeId edge) const {
+    VTR_ASSERT_SAFE(valid_edge_id(edge));
+    return edge_sink_nodes_[edge];
+}
+
+RRSwitchId RRGraph::edge_switch(RREdgeId edge) const {
+    VTR_ASSERT_SAFE(valid_edge_id(edge));
+    return edge_switches_[edge];
+}
+
+/* Check if the edge is a configurable edge (programmble) */
+bool RRGraph::edge_is_configurable(RREdgeId edge) const {
+    /* Make sure we have a valid edge id */
+    VTR_ASSERT_SAFE(valid_edge_id(edge));
+
+    auto iswitch = edge_switch(edge);
+
+    return switches_[iswitch].configurable();
+}
+
+/* Check if the edge is a non-configurable edge (hardwired) */
+bool RRGraph::edge_is_non_configurable(RREdgeId edge) const {
+    /* Make sure we have a valid edge id */
+    VTR_ASSERT_SAFE(valid_edge_id(edge));
+    return !edge_is_configurable(edge);
+}
+
+size_t RRGraph::switch_index(RRSwitchId switch_id) const {
+    VTR_ASSERT_SAFE(valid_switch_id(switch_id));
+    return size_t(switch_id);
+}
+
+/*
+ * Get a switch from the rr_switch list with a given id  
+ */
+const t_rr_switch_inf& RRGraph::get_switch(RRSwitchId switch_id) const {
+    VTR_ASSERT_SAFE(valid_switch_id(switch_id));
+    return switches_[switch_id];
+}
+
+size_t RRGraph::segment_index(RRSegmentId segment_id) const {
+    VTR_ASSERT_SAFE(valid_segment_id(segment_id));
+    return size_t(segment_id);
+}
+
+/*
+ * Get a segment from the segment list with a given id  
+ */
+const t_segment_inf& RRGraph::get_segment(RRSegmentId segment_id) const {
+    VTR_ASSERT_SAFE(valid_segment_id(segment_id));
+    return segments_[segment_id];
+}
+
+RREdgeId RRGraph::find_edge(RRNodeId src_node, RRNodeId sink_node) const {
+    for (auto edge : node_out_edges(src_node)) {
+        if (edge_sink_node(edge) == sink_node) {
+            return edge;
+        }
+    }
+
+    //Not found
+    return RREdgeId::INVALID();
+}
+
+RRNodeId RRGraph::find_node(short x, short y, t_rr_type type, int ptc, e_side side) const {
+    if (!valid_fast_node_lookup()) {
+        build_fast_node_lookup();
+    }
+    size_t itype = type;
+    size_t iside = side;
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((x < 0)                                     /* See if x is smaller than the index of first element */
+        || (size_t(x) > node_lookup_.size() - 1)) { /* See if x is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
```

Comment:
RRNodeID::INVALID() and similarly below.

Reply:
Done